Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
-			mobiledockClients[client].undockBase = Universe::get_base(iBaseID);
-			mobiledockClients[client].baseUndock = true;
+			WritePrivateProfileString("Player", "system", defaultReturnSystem.c_str(), charpath.c_str());
+			WritePrivateProfileString("Player", "base", defaultReturnBase.c_str(), charpath.c_str());
+		}
+	}
+}
 
-			return;
+bool RemoveShipFromLists(const wstring& dockedShipName, boolean isForced)
+{
+	if (!nameToDockedInfoMap.count(dockedShipName)) {
+		return false;
+	}
+	uint dockedClientID = HkGetClientIdFromCharname(dockedShipName);
 
+	if (dockedClientID != -1)
+	{
+		if (isForced)
+		{
+			ForceLaunch(dockedClientID);
+			return true;
+		}
+		idToDockedInfoMap.erase(dockedClientID);
+	}
+	else
+	{
+		//player offline, edit their character file to put them on last docked solar
+		CAccount *acc = HkGetAccountByCharname(dockedShipName);
+		if (acc)
+		{
+			MoveOfflineShipToLastDockedSolar(dockedShipName);
 		}
+	}
 
-		//Get the carrier ship information
-		uint carrierShip;
-		pub::Player::GetShip(carrier_client, carrierShip);
+	wstring& carrierName = nameToDockedInfoMap[dockedShipName].carrierName;
 
-		uint clientShip;
-		pub::Player::GetShip(client, clientShip);
+	//remove the docked ship entry from carrier's data
+	if (nameToCarrierInfoMap.count(carrierName))
+	{
+		auto& dockedList = nameToCarrierInfoMap[carrierName].dockedShipList;
+		for (auto& iter = dockedList.begin(); iter != dockedList.end(); iter++) {
+			if (*iter == dockedShipName) {
+				dockedList.erase(iter);
+				break;
+			}
+		}
+	}
+	//finally, clear the docked ship info
+	nameToDockedInfoMap.erase(dockedShipName);
 
-		// Check to see if the carrier is currently in a base. If so, force the client to dock on that base.
-		if (!carrierShip)
-		{
+	return true;
+}
 
-			uint iBaseID;
-			pub::Player::GetBase(carrier_client, iBaseID);
+// Returns count of installed docking modules on ship of specific client.
+uint GetInstalledModules(uint iClientID)
+{
+	uint modules = 0;
 
-			if (iBaseID)
+	// Check to see if the vessel undocking currently has a docking module equipped
+	for (list<EquipDesc>::iterator item = Players[iClientID].equipDescList.equip.begin(); item != Players[iClientID].equipDescList.equip.end(); item++)
+	{
+		if (find(dockingModuleEquipmentIds.begin(), dockingModuleEquipmentIds.end(), item->iArchID) != dockingModuleEquipmentIds.end())
+		{
+			if (item->bMounted)
 			{
-				// Set the flags which the PlayerLaunch_AFTER uses to handle teleporting to the carriers base
-				mobiledockClients[client].undockBase = Universe::get_base(iBaseID);
-				mobiledockClients[carrier_client].mapDockedShips.erase(clientName);
-				mobiledockClients[client].baseUndock = true;
-				return;
+				modules++;
 			}
 		}
+	}
+
+	return modules;
+}
 
-		// If carrier and docked ship are in same system.
-		if (Players[client].iSystemID == Players[carrier_client].iSystemID)
+void StartDockingProcedure(uint dockingID, uint carrierID) {
+
+	if (dockingPeriod)

```

Comment:
I must have missed this. Where is this defined, and is it configurable

Reply:
line 26, has a comment attached